pkgs/build-support/dotnet/build-dotnet-module/default.nix

```diff
@@ -30,9 +33,16 @@
 , packNupkg ? false
 # The packages project file, which contains instructions on how to compile it. This can be an array of multiple project files as well.
 , projectFile ? null
-# The NuGet dependency file. This locks all NuGet dependency versions, as otherwise they cannot be deterministically fetched.
-# This can be generated by running the `passthru.fetch-deps` script.
+
+# The NuGet dependency file. This locks all NuGet dependency versions, in case they cannot be deterministically fetched.
+# This can be generated by running the `passthru.fetch-deps` script. Note that this option is mutually exclusive with the `nugetSha256` attribute.
 , nugetDeps ? null
+# The hash of the generated nix-based lockfile containing all Nuget dependencies. This depends on no floating version ranges being requested upstream, as we cannot deterministically fetch those.
+# If any floating version is requested, an error will be thrown. Note that this option is mutually exclusive with the `nugetDeps` attribute.
+, nugetSha256 ? null
```

Comment:
What happens is the src gets updated, and the user forgets to bump nugetSha256, assuming the new update has a different set of dependencies?

Reply:
Wouldn't it just naturally fail since the hashes don't match?

Reply:
No. If you forget to update your hash you won't notice if you still have the old dependency derivation cached. It will only become visible once someone else tries to build it.

Reply:
> If you forget to update your hash you won't notice if you still have the old dependency derivation cached.

So wouldn't it fail to build?

Reply:
fixed output derivations are not invalidated. We would either need to copy the rust method or change name.

Reply:
> fixed output derivations are not invalidated.

but wouldn't the package still fail to build if the old dependency FOD is used (and is incompatible with the new src)?

Reply:
It fails to build (maybe) but having a clear error message that the hash sum needs an update is better UX than waiting for an build error.

Reply:
The way we do that in Rust is copy the Cargo.lock as part of the output of the FOD. Then in the parent derivation, we compare the $src Cargo.lock with the FOD Cargo.lock and emit an error saying that the sha256 is out of date if they don't match. It's the best way we found to emit an actionable error message.

Reply:
> What happens is the src gets updated, and the user forgets to bump nugetSha256, assuming the new update has a different set of dependencies?

Good catch, I hadn't thought of that. Most dotnet projects (in my experience) are very relaxed with dependency ranges, I've already had a few cases already where mismatching dep versions weren't caught at compile time but did break at runtime. Not a great debugging experience.

The way rust handles this seems great, but im not sure if we can adopt it for dotnet. Microsoft recommands to not check the lockfile into source control, no application currently in nixpkgs includes it as far as I'm aware. 

We could of course generate one at compile time and check against that, but there's a few downsides. To generate a lockfile you have to run `dotnet restore` which evaluates the entire dependency graph, that can take quite a while depending on the projects size. On my computer it's the main bottleneck for most packages.

We can cache this by pointing dotnet to a lockfile generated in the FOD, and running restore with `--locked-mode` when building the parent derivation. That way we only have to evaluate the dependency graph once, and cache the result. I don't think you can generate a lockfile with this flag enabled though, so we'd have to run `dotnet restore` twice and would obviously only be able catch a hash mismatch mid compilation.

The only way I can think of to catch this at evaluation time is to make the FOD dependant on `name` like @SuperSandro2000 suggested. Unless someone has a better idea I'll go with that :+1: